plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2021,24 +2136,40 @@ producer_plugin_impl::push_transaction( const fc::time_point& block_deadline,
       }
    }
 
-   // setting and unsetting chainbase read_only mode will be moved to right
-   // place when multi-threaded read-only transaction is implemented
-   auto unset_db_read_only_mode = fc::make_scoped_exit([trx, &chain]{
+   // Make sure db_read_only_mode to be always unset
+   auto db_read_only_mode_guard = fc::make_scoped_exit([trx, &chain]{
       if( trx->is_read_only() )
          chain.unset_db_read_only_mode();
    });
    if( trx->is_read_only() )
       chain.set_db_read_only_mode();
+
    auto trace = chain.push_transaction( trx, block_deadline, max_trx_time, prev_billed_cpu_time_us, false, sub_bill );
+
    if( trx->is_read_only() ) {
       chain.unset_db_read_only_mode();
-      unset_db_read_only_mode.cancel();
+      db_read_only_mode_guard.cancel();
    }
 
+   return handle_push_result(trx, next, start, chain, trace, return_failure_trace, disable_subjective_enforcement, first_auth, sub_bill, prev_billed_cpu_time_us);
+}
+
+producer_plugin_impl::push_result
+producer_plugin_impl::handle_push_result( const transaction_metadata_ptr& trx,
+                                          next_function<transaction_trace_ptr> next,
+                                          const fc::time_point& start,
+                                          const chain::controller& chain,
+                                          const transaction_trace_ptr& trace,
+                                          bool return_failure_trace,
+                                          bool disable_subjective_enforcement,
+                                          account_name first_auth,
+                                          int64_t sub_bill,
+                                          uint32_t prev_billed_cpu_time_us) {
    auto end = fc::time_point::now();
    push_result pr;
    if( trace->except ) {
-      _account_fails.add_fail_time(end - start);
+      if ( !trx->is_read_only() )
+         _account_fails.add_fail_time(end - start);
       if( exception_is_exhausted( *trace->except ) ) {
```

Comment:
You could skip the `block_is_exhausted()` call for read-only as it should never be exhausted since you have disabled running on a producing node.

Reply:
Thanks.